### PR TITLE
Followup of #11095

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

This PR is followup of #11095 to address comments from @mtreinish along with a bugfix. The bug is reported in Qiskit Experiments, in which backend version up-conversion fails when the value of gate config is invalid. This is an edge case of the [FakeOpenPulse2Q](https://github.com/Qiskit/qiskit/blob/7a411405ac3b5f2e107eabb8846475c38fdcaa07/qiskit/providers/fake_provider/fake_openpulse_2q.py#L47) backend, because its gate config doesn't match with the basis gates that it reports. However this must be resolved.

### Details and comments

Code cleanup + Bugfix

Several cleanup is done in this PR.
- Typehint fix
- Handling of faulty qubits
- Direct instantiation of instruction properties object
- More inline comments for future developer

### CODE REVIEW

File: qiskit/providers/backend_compat.py [hunk 1]
```diff
@@ -119,29 +121,43 @@ def convert_to_target(
...
 
-    # Create empty inst properties from gate configs
-    for name, spec in gate_configs.items():
-        if hasattr(spec, "coupling_map"):
-            coupling_map = spec.coupling_map
-            prop_name_map[name] = dict.fromkeys(map(tuple, coupling_map))
-        else:
+    # Create inst properties placeholder
+    for name in all_instructions:
+        if name not in prop_name_map:
```
Comment (mtreinish) [inline qiskit/providers/backend_compat.py]: Can this ever be `True`? `all_instructions` is a set so there aren't any duplicates and `prop_name_map` on the first iteration is empty.

File: qiskit/providers/backend_compat.py [hunk 2]
```diff
@@ -153,32 +169,26 @@ def convert_to_target(
...
+                        set.intersection(faulty_qubits, qubits)
+                        or not properties.is_gate_operational(name, qubits)
+                    ):
+                        faulty_ops.add((name, qubits))
+                        continue
+                    if prop_name_map[name] is None:
+                        # This instruction is tied to particular qubits
+                        # i.e. gate config is not provided, and instruction has been globally defined.
+                        prop_name_map[name] = {}
+                    prop_name_map[name][qubits] = InstructionProperties(
```
Comment (mtreinish) [inline qiskit/providers/backend_compat.py:191]: Are you worried about a potential conflict here between the `BackendProperties` and the gate config above? I could rant about the poor interface for the v1 model objects, but if there is a mismatch here you'll potentially end up with a properties dict like:

```
{
  (0,): None,
  (1,): InstructionProperties(...),
  (3,): InstructionProperties(...),
}
```

Which I guess would work fine but is a bit weird to say the gate is an ideal on qubit 0 and has properties on 1 and 3. TBH, I don't remember how the converter handled inconsistent data in a v1 backend before so this is probably fine.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM it's a nice cleanup, just a couple small questions inline

File: qiskit/providers/backend_compat.py (hunk 1 above)
Comment (nkanazawa1989) [reply qiskit/providers/backend_compat.py]: You're right. Updated initialization of `prop_name_map`

File: qiskit/providers/backend_compat.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/providers/backend_compat.py:191]: I think providing correct gate config is responsibility of the provider. I just added logger message.

**Review by mtreinish — COMMENTED** [review]
This LGTM, my only question before approving is around the release notes from #11095. Did we add the upgrade notes about the api changes, see:

https://github.com/Qiskit/qiskit/pull/11095#discussion_r1422425887 and https://github.com/Qiskit/qiskit/pull/11095#discussion_r1422426663 (they might have been done in another PR and I just forgot during my winter vacation)

**Comment by nkanazawa1989** [comment]
Good catch! Updated the reno in [5f7bc6e](https://github.com/Qiskit/qiskit/pull/11404/commits/5f7bc6e5a49d836f042cff67e08a36e461a2ab95) (also rephrased a bit).

File: qiskit/providers/backend_compat.py [hunk 3]
```diff
@@ -187,18 +214,17 @@ def convert_to_target(
...
 
-    if add_delay and "delay" not in prop_name_map:
-        prop_name_map["delay"] = {
-            (q,): None for q in range(configuration.num_qubits) if q not in faulty_qubits
-        }
+    for op in required:
+        # Map required ops to each operational qubit
+        if prop_name_map[op] is None:
+            prop_name_map[op] = {
+                (q,): None for q in range(configuration.num_qubits) if q not in faulty_qubits
```
Comment (mtreinish) [inline qiskit/providers/backend_compat.py]: Shouldn't we only exclude the faulty qubits here if `filter_faulty` is `True`?

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, just one tiny issue inline in the logic around adding required instructions to the target.

File: qiskit/providers/backend_compat.py (hunk 3 above)
Comment (nkanazawa1989) [reply qiskit/providers/backend_compat.py]: Done in [bcb4144](https://github.com/Qiskit/qiskit/pull/11404/commits/bcb4144567c2ea92714cc01e5241b2b7e9c3f557) (also added a test).

**Review by mtreinish — APPROVED** [review]
LGTM now, thanks for the quick update

**Pull request merged**